activerecord/lib/active_record/relation/delegation.rb

```diff
@@ -61,10 +61,13 @@ def generate_method(method)
 
           if /\A[a-zA-Z_]\w*[!?]?\z/.match?(method)
             definition = RUBY_VERSION >= "2.7" ? "..." : "*args, &block"
+            temp_name = "__temp__#{method}"
```

Comment:
What's the overhead of `def / alias / undef`? 

Because if it's important, it might be better to simply fallback to `define_method` for keyword names.

Reply:
What you mean is like here?

```diff
diff --git a/activerecord/lib/active_record/relation/delegation.rb b/activerecord/lib/active_record/relation/delegation.rb
index 6c809adf3b..19b7c46573 100644
--- a/activerecord/lib/active_record/relation/delegation.rb
+++ b/activerecord/lib/active_record/relation/delegation.rb
@@ -1,6 +1,7 @@
 # frozen_string_literal: true
 
 require "mutex_m"
+require "active_support/core_ext/module/delegation"
 
 module ActiveRecord
   module Delegation # :nodoc:
@@ -59,7 +60,7 @@ def generate_method(method)
         synchronize do
           return if method_defined?(method)
 
-          if /\A[a-zA-Z_]\w*[!?]?\z/.match?(method)
+          if /\A[a-zA-Z_]\w*[!?]?\z/.match?(method) && !DELEGATION_RESERVED_METHOD_NAMES.include?(method)
             definition = RUBY_VERSION >= "2.7" ? "..." : "*args, &block"
             temp_name = "__temp__#{method}"
             module_eval <<-RUBY, __FILE__, __LINE__ + 1
```

I'm not sure the overhead, what I know why don't `define_method` for all is, defined method by `define_method` is slow.

Reply:
> you mean is like here

Yes.

I know `define_method` is slower on execution, so I'm indeed not suggesting to always use it. However I think it's simpler code to use it for ruby keywords, it avoid compelxifying the happy path.

Reply:
I see, I've changed the direction in 9927d8f368a2c8061f557b5814776c58a2ef219d.